types/module.go

```diff
@@ -0,0 +1,194 @@
+	GetTxCmd() *cobra.Command
+}
+
+// AppModule is the standard form for an application module
+type AppModule interface {
+
+	// app name
+	Name() string
+
+	// registers
+	RegisterInvariants(InvariantRouter)
+
+	// routes
+	Route() string
+	NewHandler() Handler
+	QuerierRoute() string
+	NewQuerierHandler() Querier
+
+	//// genesis
+	//DefaultGenesisState() json.RawMessage
+	//ValidateGenesis(json.RawMessage) error
+	InitGenesis(Context, json.RawMessage) ([]abci.ValidatorUpdate, error)
+	//ExportGenesis(Context) json.RawMessage
+
+	BeginBlock(Context, abci.RequestBeginBlock) Tags
+	EndBlock(Context, abci.RequestEndBlock) ([]abci.ValidatorUpdate, Tags)
+}
+
+// module manananaager
+type ModuleManager struct {
+	Modules            map[string]AppModule
+	OrderInitGenesis   []string
+	OrderExportGenesis []string
+	OrderBeginBlockers []string
+	OrderEndBlockers   []string
+}
+
+// NewModuleManager creates a new ModuleManager object
+func NewModuleManager(modules ...AppModule) *ModuleManager {
+
+	moduleMap := make(map[string]AppModule)
+	var modulesStr []string
+	for _, module := range modules {
+		moduleMap[module.Name()] = module
+		modulesStr = append(modulesStr, module.Name())
+	}
+
+	return &ModuleManager{
+		Modules:            moduleMap,
+		OrderInitGenesis:   modulesStr,
+		OrderExportGenesis: modulesStr,
+		OrderBeginBlockers: modulesStr,
+		OrderEndBlockers:   modulesStr,
+	}
+}
+
+// set the order of init genesis calls
+func (mm *ModuleManager) SetOrderInitGenesis(moduleNames ...string) {
+	mm.OrderInitGenesis = moduleNames
+}
+
+// set the order of export genesis calls
+func (mm *ModuleManager) SetOrderExportGenesis(moduleNames ...string) {
+	mm.OrderExportGenesis = moduleNames
+}
+
+// set the order of set begin-blocker calls
+func (mm *ModuleManager) SetOrderBeginBlockers(moduleNames ...string) {
+	mm.OrderBeginBlockers = moduleNames
+}
+
+// set the order of set end-blocker calls
+func (mm *ModuleManager) SetOrderEndBlockers(moduleNames ...string) {
+	mm.OrderEndBlockers = moduleNames
+}
+
+// register all module routes and module querier routes
+func (mm *ModuleManager) RegisterInvariants(invarRouter InvariantRouter) {
+	for _, module := range mm.Modules {
+		module.RegisterInvariants(invarRouter)
+	}
+}
+
+// register all module routes and module querier routes
+func (mm *ModuleManager) RegisterRoutes(router Router, queryRouter QueryRouter) {
+	for _, module := range mm.Modules {
+		if module.Route() != "" {
+			router.AddRoute(module.Route(), module.NewHandler())
+		}
+		if module.QuerierRoute() != "" {
+			queryRouter.AddRoute(module.QuerierRoute(), module.NewQuerierHandler())
+		}
+	}
+}
+
+//// validate all genesis information
+//func (mm *ModuleManager) ValidateGenesis(genesisData map[string]json.RawMessage) error {
+//for _, module := range mm.Modules {
+//err := module.ValidateGenesis(genesisDate[module.Name()])
+//if err != nil {
+//return err
+//}
+//}
+//return nil
+//}
+
+//// default genesis state for modules
```

Comment:
I guess all commented line would be deleted, yeah?

Reply:
These are reworked/uncommented within the next PR